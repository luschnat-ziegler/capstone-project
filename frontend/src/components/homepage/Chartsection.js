import styled from 'styled-components/macro'
import { scaleSequential } from 'd3-scale'
import { interpolateRdYlGn } from 'd3-scale-chromatic'
import PropTypes from 'prop-types'

export default function Chartsection({ displayedCountries, entry, heading }) {
  return (
    <>
      <CategoryHeading>{heading}</CategoryHeading>
      {displayedCountries.countryLeft.hasOwnProperty('country') &&
        (displayedCountries.countryLeft[entry] ? (
          <Bar data={displayedCountries.countryLeft[entry]} left={true} />
        ) : (
          <NoDataBar left={true}>No data</NoDataBar>
        ))}
      <Divider />
      {displayedCountries.countryRight.hasOwnProperty('country') &&
        (displayedCountries.countryRight[entry] ? (
          <Bar data={displayedCountries.countryRight[entry]} left={false} />
        ) : (
          <NoDataBar left={false}>No data</NoDataBar>
        ))}
    </>
  )
}

const colorScale = scaleSequential().domain([0, 100]).interpolator(interpolateRdYlGn)

const CategoryHeading = styled.h3`
  grid-column: 1 / span 3;
  padding: 10px;
  color: #404040;
`

const Divider = styled.div`
  background-color: #e8eddf;
  border: 1px solid grey;
  grid-column: 2 / span 1;
  height: 30px;
  width: 100%;
`

const Bar = styled.div`
  border: 1px solid grey;
  border-right: ${(props) => props.left && 'none'};
  border-left: ${(props) => !props.left && 'none'};
  border-radius: ${(props) => (props.left ? '5px 0 0 5px' : '0 5px 5px 0')};
  background-color: ${(props) => colorScale(props.data)};
  height: 30px;
  width: ${(props) => String(props.data * 0.9) + '%'};
  justify-self: ${(props) => props.left && 'end'};
`

const NoDataBar = styled.div`
  padding: 5px;
  background-color: whitesmoke;
  color: grey;
  border: 1px solid grey;
  border-style: dotted;
  border-right: ${(props) => props.left && 'none'};
  border-left: ${(props) => !props.left && 'none'};
  border-radius: ${(props) => (props.left ? '5px 0 0 5px' : '0 5px 5px 0')};
  height: 30px;
  width: 50%;
  justify-self: ${(props) => props.left && 'end'};
`

Chartsection.propTypes = {
  displayedCountries: PropTypes.object,
  entry: PropTypes.string,
  heading: PropTypes.string,
}

Bar.propTypes = {
  data: PropTypes.number,
  left: PropTypes.bool,
}

NoDataBar.propTypes = {
  left: PropTypes.bool,
}
