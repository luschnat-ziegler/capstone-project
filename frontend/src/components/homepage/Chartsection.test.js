import { render } from '@testing-library/react'
import 'jest-styled-components'
import Chartsection from './Chartsection'

const displayedCountries = {
  countryLeft: {
    id: 18,
    country: 'Bhutan',
    region: 'Asia',
    freedom: 59,
    gender: 49,
    lgbtq: 35,
    environment: null,
    corruption: 75,
    inequality: 65,
    total: 57,
  },
  countryRight: {
    id: 4,
    country: 'Angola',
    region: 'SSA',
    freedom: 32,
    gender: 32,
    lgbtq: 29,
    environment: 89,
    corruption: 19,
    inequality: 30,
    total: 38,
  },
}

describe('Chartsection', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Chartsection
        displayedCountries={displayedCountries}
        heading={'Corruption Control'}
        entry={'corruption'}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('has the correct heading', () => {
    const { getByRole } = render(
      <Chartsection
        displayedCountries={displayedCountries}
        heading={'Corruption Control'}
        entry={'corruption'}
      />
    )

    const heading = getByRole('heading')
    expect(heading).toHaveTextContent(/corruption/i)
  })

  it('has a divider div', () => {
    const { container } = render(
      <Chartsection
        displayedCountries={displayedCountries}
        heading={'Corruption Control'}
        entry={'corruption'}
      />
    )

    const divider = container.children[2]
    expect(divider).toBeInTheDocument()
    expect(divider).toHaveStyleRule('background-color', '#e8eddf')
    expect(divider).toHaveStyleRule('grid-column', '2 / span 1')
  })

  it('has two bars of correct length', () => {
    const { container } = render(
      <Chartsection
        displayedCountries={displayedCountries}
        heading={'Corruption Control'}
        entry={'corruption'}
      />
    )

    const leftBar = container.children[1]
    const rightBar = container.children[3]
    expect(leftBar).toBeInTheDocument()
    expect(rightBar).toBeInTheDocument()
    expect(leftBar).toHaveStyleRule('width', '67.5%')
    expect(rightBar).toHaveStyleRule('width', '17.1%')
  })
})
