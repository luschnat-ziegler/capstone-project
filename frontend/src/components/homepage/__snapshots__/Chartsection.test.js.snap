// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chartsection renders correctly 1`] = `
.c0 {
  grid-column: 1 / span 3;
  padding: 10px;
  color: #404040;
}

.c2 {
  background-color: #e8eddf;
  border: 1px solid grey;
  grid-column: 2 / span 1;
  height: 30px;
  width: 100%;
}

.c1 {
  border: 1px solid grey;
  border-right: none;
  border-radius: 5px 0 0 5px;
  background-color: rgb(133,203,103);
  height: 30px;
  width: 67.5%;
  justify-self: end;
}

.c3 {
  border: 1px solid grey;
  border-left: none;
  border-radius: 0 5px 5px 0;
  background-color: rgb(239,103,64);
  height: 30px;
  width: 17.1%;
}

<div>
  <h3
    class="c0"
  >
    Corruption Control
  </h3>
  <div
    class="c1"
    data="75"
  />
  <div
    class="c2"
  />
  <div
    class="c3"
    data="19"
  />
</div>
`;
